src/irmin/tree_intf.ml

```diff
@@ -385,5 +386,16 @@ module type Tree = sig
     val hash : ?cache:bool -> t -> kinded_hash
     val of_private_node : P.Repo.t -> P.Node.value -> node
     val to_private_node : node -> P.Node.value or_error Lwt.t
+
+    (** {2 Proofs} *)
+
+    type proof =
+      [ `Tree of P.Node.Val.proof * (step * proof) list
```

Comment:
we should probably inline node proofs here, as we don't need to store the step twice and intermediate hashes.